crates/ruff_python_ast/src/nodes.rs

```diff
@@ -767,12 +767,62 @@ impl From<ExprIf> for Expr {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct DictItem {
+    pub key: Option<Expr>,
+    pub value: Expr,
+}
+
+impl DictItem {
+    fn key(&self) -> Option<&Expr> {
+        self.key.as_ref()
+    }
+
+    fn value(&self) -> &Expr {
+        &self.value
+    }
+}
+
+impl Ranged for DictItem {
+    fn range(&self) -> TextRange {
+        TextRange::new(
+            self.key.as_ref().map_or(self.value.start(), Ranged::start),
+            self.value.end(),
+        )
+    }
+}
+
 /// See also [Dict](https://docs.python.org/3/library/ast.html#ast.Dict)
 #[derive(Clone, Debug, PartialEq)]
 pub struct ExprDict {
     pub range: TextRange,
-    pub keys: Vec<Option<Expr>>,
-    pub values: Vec<Expr>,
+    pub items: Vec<DictItem>,
+}
+
+impl ExprDict {
+    pub fn iter_keys(&self) -> DictKeyIterator {
+        DictKeyIterator::new(&self.items)
+    }
+
+    pub fn iter_values(&self) -> DictValueIterator {
+        DictValueIterator::new(&self.items)
+    }
+
+    pub fn iter_items(&self) -> Iter<'_, DictItem> {
+        self.items.iter()
+    }
+
+    pub fn key(&self, n: usize) -> Option<&Expr> {
+        self.items[n].key()
+    }
+
+    pub fn value(&self, n: usize) -> &Expr {
+        self.items[n].value()
+    }
+
+    pub fn item(&self, n: usize) -> &DictItem {
+        &self.items[n]
+    }
 }
```

Comment:
I think we could implement `Index` for `item`, if that's even something that's commonly used.

Reply:
I'm not sure it's a common enough need for it to be worth it. But if we did do this, I think (same as https://github.com/astral-sh/ruff/pull/11267#discussion_r1591893793) we should do it for `ExprList`, `ExprSet` and `ExprTuple` at the same time, same as https://github.com/astral-sh/ruff/pull/11267#discussion_r1591893793